File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <bits/stdc++.h>
using namespace std;

int dp[500][500], res;

int solution(vector<vector<int>> triangle) {
    int N = triangle.size();
    dp[0][0] = triangle[0][0];
    for (int i = 0; i < N - 1; i++)
        for (int j = 0; j <= i; j++) {
            dp[i + 1][j] = max(dp[i + 1][j], dp[i][j] + triangle[i + 1][j]);
            dp[i + 1][j + 1] = max(dp[i + 1][j + 1], dp[i][j] + triangle[i + 1][j + 1]);
        }
    for (int i = 0; i < N; i++)
        res = max(res, dp[N - 1][i]);
    return res;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int N = triangle.length;
        int[][] cache = new int[N][N];
        cache[0][0] = triangle[0][0];
        
        for (int i = 0; i < N - 1; i++)
            for (int j = 0; j <= i; j++) {
                cache[i + 1][j] = max(cache[i + 1][j], 
                                      cache[i][j] + triangle[i + 1][j]);
                cache[i + 1][j + 1] = max(cache[i + 1][j + 1], 
                                          cache[i][j] + triangle[i + 1][j + 1]);
            }
        
        int res = 0;
        for (int j = 0; j < N; j++)
            res = max(res, cache[N - 1][j]);
        return res;
    }
    
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
}